### PR TITLE
Class based Optimize solvers.

### PR DESCRIPTION
**Pull request by andyfaff**

### CODE REVIEW

**Comment by andyfaff** [comment]
@stsievert, so one can see the CI.

File: scipy/optimize/optimize.py [hunk 1]
```diff
@@ -3061,6 +2904,580 @@ def show_options(solver=None, method=None, disp=True):
...
+    Optionally, the lower and upper bounds for each element in x can also be
+    specified using the `bounds` argument.
+    """
+    opt_options = {'args': (), 'bounds': None, 'x0': None, 'jac': None,
+                   'hess': None, 'maxiter': np.inf, 'maxfun': np.inf, 'disp': False,
+                   'callback': None}
+
+    def __init__(self, func, **options):
+        self.opt_options.update(options)
+        if isinstance(func, OptimizeFunction):
```
Comment (stsievert) [inline scipy/optimize/optimize.py]: Do we want to issue a deprecation warning here if `args`, `jac` or `hess` specified?

Either way we should mention the API change in the docstring.

File: scipy/optimize/optimize.py [hunk 2]
```diff
@@ -3061,6 +2904,580 @@ def show_options(solver=None, method=None, disp=True):
...
+    @property
+    def N(self):
+        """
+        The dimensionality of the problem, `np.size(x)`.
+        """
+        if self.x is not None:
+            return self.x.size
+        elif self.x0 is not None:
+            return self.x0.size
+        else:
```
Comment (stsievert) [inline scipy/optimize/optimize.py]: The `return`s make the `else` unnecessary

File: scipy/optimize/optimize.py [hunk 3]
```diff
@@ -3061,6 +2904,580 @@ def show_options(solver=None, method=None, disp=True):
...
+        """
+        # TODO interupt once convergence is reached?
+        for it in range(iterations):
+            self.x, self.fun = next(self)
+
+        self._finish_up()
+        return self.result
+
+    @property
+    def N(self):
```
Comment (stsievert) [inline scipy/optimize/optimize.py]: I like the name `dim` over single letter names (like `n`, `p` or `d`).

File: scipy/optimize/optimize.py [hunk 4]
```diff
@@ -3061,6 +2904,580 @@ def show_options(solver=None, method=None, disp=True):
...
+        methods, and during `__exit__` of the context manager.
+        """
+        pass
+
+    def show_options(self):
+        # each optimizer should return their own options text.
+        return ""
+
+
+class OptimizeFunction(object):
```
Comment (stsievert) [inline scipy/optimize/optimize.py]: Could we name it `Function` instead? It's already inside scipy.optimize.

File: scipy/optimize/optimize.py (hunk 1 above)
Comment (andyfaff) [reply scipy/optimize/optimize.py]: Removed jac, args and hess. This is a new interface, better to be clean to start with. People will still be able to use the old `minimize` interface for a long time yet.

File: scipy/optimize/optimize.py [hunk 5]
```diff
@@ -3061,6 +2912,576 @@ def show_options(solver=None, method=None, disp=True):
...
+
+    def show_options(self):
+        # each optimizer should return their own options text.
+        return ""
+
+
+class Function(object):
+    def __init__(self, func=None, args=(), jac=None, hess=None):
+        self._func = func
+        self.args = args
```
Comment (stsievert) [inline scipy/optimize/optimize.py]: Do we still want `args`? I remember it being a pain point. I don't think we want to remove it for backwards compatibility etc.

File: scipy/optimize/optimize.py [hunk 6]
```diff
@@ -3061,6 +2912,576 @@ def show_options(solver=None, method=None, disp=True):
...
+
+    def __next__(self):
+        """
+        Advance the solver by a single iteration.
+        """
+        # Should be over-ridden by each class based solver.
+        # TODO, Stop if self.maxfun/self.maxiter reached?
+        raise NotImplementedError
+    # step is because PyTorch has a step method
+    next = step = __next__
```
Comment (stsievert) [inline scipy/optimize/optimize.py]: We should move this to `__init__` and have `self.next = self.step = self.__next__`. That makes sure they all refer to the same object (at least when `super().__init__` called)

File: scipy/optimize/optimize.py [hunk 7]
```diff
@@ -3061,6 +2912,576 @@ def show_options(solver=None, method=None, disp=True):
...
+            self.x, self.fun = x, fun
+
+            if (self.converged() or self.nfev >= self.maxfun or
+                    self.nit >= self.maxiter):
+                break
+
+        self._finish_up()
+        return self.result
+
+    def converged(self):
```
Comment (stsievert) [inline scipy/optimize/optimize.py]: Maybe we should implement this. We pass `maxiter` and  `maxfun` to the Optimizer; should we use them?

**Comment by stsievert** [comment]
This is looking pretty good (and way better than my implementation, [objopt]). It looks like the interface I'd like is possible:

[objopt]:https://github.com/stsievert/objopt

``` python
class L2Loss(Function):
    def __init__(self, A, y):
        self.A = A
        self.y = y

    def func(x):
        return LA.norm(self.A@x - self.y)**2

    def grad(x):
        return 2 * self.A.T @ (self.A@x - self.y)

if __name__ == "__main__":
    n, d = 100, 10
    A = np.random.randn(n, d)
    x_star = np.random.randn(d)
    y = np.sign(A @ x)

    loss = L2Loss(A, y)
    opt = Optimizer(loss)

    for iteration in range(100):
        opt.step()
        # use opt to compute other stats
    # use opt to calculate more stats
```

File: scipy/optimize/optimize.py (hunk 5 above)
Comment (stsievert) [reply scipy/optimize/optimize.py]: Could we have some keyword arguments too?

Comment (andyfaff) [reply scipy/optimize/optimize.py]: I don't see why not.

Comment (andyfaff) [reply scipy/optimize/optimize.py]: Done

File: scipy/optimize/optimize.py (hunk 6 above)
Comment (andyfaff) [reply scipy/optimize/optimize.py]: Done

File: scipy/optimize/optimize.py (hunk 7 above)
Comment (andyfaff) [reply scipy/optimize/optimize.py]: Hmm. Personally I don't think `self.nit >=self.maxiter` belong in `converged`. For example, you may set up a maxiter that's too low to reach convergence, but then still want to do more iterations.

**Comment by andyfaff** [comment]
BTW, if I miss a message it's because I'm a bit overwhelmed with features to implement, so may need reminding.

File: scipy/optimize/optimize.py [hunk 8]
```diff
@@ -3061,6 +2912,567 @@ def show_options(solver=None, method=None, disp=True):
...
+        self.nhev = 0
+
+    def _same_x(self, x):
+        return numpy.array_equal(x, self._last_x)
+
+    # TODO thought of using decorators to count calls of func, jac, hess,
+    # but overridden methods would probably forget to do so.
+    def __call__(self, x):
+        # do not override this method otherwise nfev may not be correct
+        self.nfev += 1
```
Comment (jaimefrio) [inline scipy/optimize/optimize.py]: Having state in the Function object is dangerous, because it is tempting to reuse it. Leaving aside the exact actual way to advance the iteration, you may want to do something like:

```
f = Function(my_func, jac=my_grad)

with NelderMead(f, x0, , maxfun=100) as opt:
    while opt.step() and not opt.converged():
        print('Still going...')

with LBFGSB(f, x0, maxfun=100) as opt:
    while opt.step() and not opt.converged():
        print('Still going...')

```

And the second time around you'd start counting function evaluations where the first optimization finished, which is probably not what you want to do...

Comment (andyfaff) [reply scipy/optimize/optimize.py]: That's a very good point. Not sure how to proceed on this one. I tried using decorators but that won't work with an overridden method. Any suggestuona.?

Comment (jaimefrio) [reply scipy/optimize/optimize.py]: Not really... The obvious solution seems to be to keep all the state in one place. I'd say in the Optimizer? It would make sense especially if it operates as a context manager, but I'm not sure how that could work. One can also think of making Function a context manager, and have a factory function so that you could do:

```
f = Function(...)

with f.reset_all() as f_:
    # f_ is a copy of f with all evaluations reset
```

But frankly, it is a little cumbersome...

I need to find the time to review your code in depth and think this whole thing through, but I wanted to leave a note here so I didn't forget.

File: scipy/optimize/optimize.py (hunk 7 above)
Comment (stsievert) [reply scipy/optimize/optimize.py]: > Personally I don't think `self.nit >=self.maxiter` belong in `converged`

Fair point. Then I think it's best we leave it unimplemented. We could define another function, but I'd rather not have the overhead.

File: scipy/optimize/optimize.py (hunk 8 above)
Comment (stsievert) [reply scipy/optimize/optimize.py]: But storing the number of calls to the function/jacobian/hessian in `Function` is natural and where I'd expect to find these values.

We could have the optimizer reset the state of the function when initialized.

File: scipy/optimize/optimize.py (hunk 7 above)
Comment (andyfaff) [reply scipy/optimize/optimize.py]: After some changes the only place where maxiter and maxfun reside are in the solve method.

Comment (stsievert) [reply scipy/optimize/optimize.py]: I think that's fair. I'd rather provide a framework and let the user decide.

**Comment by stsievert** [comment]
I'm having Git difficulties, and want to propose a change. We can still support the old API with this function:

``` python
def minimize(func, x0, *args_, args=(), method=None, jac=None, hess=None, **kwargs):
    solve_kwargs = {key: kwargs.pop(key) for key in ['maxiter', 'callback',
                                                     'disp', 'maxfun']
                    if key in kwargs}
    fn = Function(func, args=args, grad=jac, hess=hess)
    fn.x = x0
    opts = [opt for opt in sys.modules[__name__]]
    viable_opts = [opt for opt in opts
                   if opt.__name__.lower() == method.lower().replace('-', '')]
    if len(viable_opts) > 1:
        raise ValueError('More than 1 viable optimizer was found for '
                         'method=' + str(method))
    opt = viable_opts[0](fn, *args_, **kwargs)
    return opt.solve(**solve_kwargs)
```

...though it's untested.
